src/NzbDrone.Core/Download/Clients/DownloadStation/Proxies/DiskStationProxyBase.cs

```diff
@@ -174,46 +202,46 @@ private HttpRequest BuildRequest(DownloadStationSettings settings, DiskStationAp
+                 { "query", string.Join(",", apis.Keys) },
              };
 
             var infoResponse = ProcessRequest<DiskStationApiInfoResponse>(DiskStationApi.Info, arguments, settings, "Get api version");
 
-            //TODO: Refactor this into more elegant code
-            var infoResponeDSAuth = infoResponse.Data["SYNO.API.Auth"];
-            var infoResponeDSInfo = infoResponse.Data["SYNO.DownloadStation.Info"];
-            var infoResponeDSTask = infoResponse.Data["SYNO.DownloadStation.Task"];
-            var infoResponseFSList = infoResponse.Data["SYNO.FileStation.List"];
-            var infoResponseDSMInfo = infoResponse.Data["SYNO.DSM.Info"];
-
-            Resources[DiskStationApi.Auth] = infoResponeDSAuth.Path;
-            Resources[DiskStationApi.DownloadStationInfo] = infoResponeDSInfo.Path;
-            Resources[DiskStationApi.DownloadStationTask] = infoResponeDSTask.Path;
-            Resources[DiskStationApi.FileStationList] = infoResponseFSList.Path;
-            Resources[DiskStationApi.DSMInfo] = infoResponseDSMInfo.Path;
+            foreach (var data in infoResponse.Data)
+            {
+                if (apis.ContainsKey(data.Key))
+                {
+                    _resources[apis[data.Key]] = data.Value;
+                }
+            }
 
-            switch (api)
+            if (_resources.ContainsKey(Api.Value))
             {
-                case DiskStationApi.Auth:
-                    return Enumerable.Range(infoResponeDSAuth.MinVersion, infoResponeDSAuth.MaxVersion - infoResponeDSAuth.MinVersion + 1);
-                case DiskStationApi.DownloadStationInfo:
-                    return Enumerable.Range(infoResponeDSInfo.MinVersion, infoResponeDSInfo.MaxVersion - infoResponeDSInfo.MinVersion + 1);
-                case DiskStationApi.DownloadStationTask:
-                    return Enumerable.Range(infoResponeDSTask.MinVersion, infoResponeDSTask.MaxVersion - infoResponeDSTask.MinVersion + 1);
-                case DiskStationApi.FileStationList:
-                    return Enumerable.Range(infoResponseFSList.MinVersion, infoResponseFSList.MaxVersion - infoResponseFSList.MinVersion + 1);
-                case DiskStationApi.DSMInfo:
-                    return Enumerable.Range(infoResponseDSMInfo.MinVersion, infoResponseDSMInfo.MaxVersion - infoResponseDSMInfo.MinVersion + 1);
-                default:
-                    throw new DownloadClientException("Api not implemented");
+                return Enumerable.Range(_resources[Api.Value].MinVersion, _resources[Api.Value].MaxVersion - _resources[Api.Value].MinVersion + 1);
             }
+
+            throw new DownloadClientException("Api not implemented");
+        }
+
+        protected int GetMinVersion(DownloadStationSettings settings)
+        {
+            return _cache.Get(Api.Key, () => GetApiVersion(settings).Min(), TimeSpan.FromHours(1));
```

Comment:
The cache key isn't unique per host:port, so you still have the 'two synology' issue.